### PR TITLE
Wip experimental udp transport

### PR DESCRIPTION
**Pull request by drewhk**
**DO NOT MERGE**

This is just a PoC, not production quality at all, missing tests, shortcuts, whatnot. Does not need a detailed review, purpose is to inspire brainstorming.

Basic items here:
- Udp based, spinning-thread driver to send/receive UDP stuff
- SPSC based communication from Flow representing transport and between driver spinning thread
- Allocation free design using preallocated Frames (slices into DirectBuffer), exploiting backpressure/boundedness
- Transparent plug-in into existing Flow (does not matter who sends first, they will share the same underlying Registration)
- Flow control and message redelivery (still buggy)

What I couldn't finish in this PoC
- handshake, UID management/quarantine
- fragmenting/defragmenting 
- subchannels

Basically this is just a bunch of ideas, many of these can be used with Aeron as well (I used plain UDP here but the style is similar).

### CODE REVIEW

File: akka-remote/src/main/scala/akka/remote/artery/driver/FlowAndLossControl.scala [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+  def apply(frameBuffer: FrameBuffer): BidiFlow[ByteString, Frame, Frame, ByteString, NotUsed] = {
+    BidiFlow.fromGraph(new FlowAndLossControlStage(frameBuffer))
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] final class FlowAndLossControlStage(sndFramePool: FrameBuffer) extends GraphStage[BidiShape[ByteString, Frame, Frame, ByteString]] {
```
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/driver/FlowAndLossControl.scala]: This class should be factored out into inbound and outbound buffers so that they encapsulate the necessary states for these directions and can be tested in isolation.

File: akka-remote/src/main/scala/akka/remote/artery/driver/UdpTransport.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+import akka.stream.scaladsl.Flow
+import akka.stream.stage._
+
+/**
+ * INTERNAL API
+ */
+private[remote] class UdpTransport(driver: UdpDriver) {
+
+  def forRemoteAddress(address: InetSocketAddress): Flow[Frame, Frame, NotUsed] = {
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/driver/UdpTransport.scala]: the part that I don't get with the `Flow` approach is how this would be created and connected for inbound messages, when the system is only receiving messages and not creating outbound associations
